pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,272 @@
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+	}
+
+	d.receivedRequests = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_received_requests_total",
+		Help: "The total number of requests received.",
+	}, []string{"user"})
+	d.amSends = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_total",
+		Help: "The total number of requests sent to the alertmanager.",
+	}, []string{"ingester"})
+	d.amSendFailures = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_failures_total",
+		Help: "The total number of requests sent to the alertmanager that failed.",
+	}, []string{"ingester"})
+
+	d.Service = services.NewBasicService(d.starting, d.running, d.stopping)
+	return d, nil
+}
+
+func (d *Distributor) starting(ctx context.Context) error {
+	return nil
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+func (d *Distributor) stopping(_ error) error {
+	return nil
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	logger := log.With(d.logger, "user", userID)
+
+	d.receivedRequests.WithLabelValues(userID).Inc()
+
+	if r.Method == http.MethodGet {
+		d.doRead(userID, w, r, logger)
+	} else {
+		d.doWrite(userID, w, r, logger)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	source := util.GetSourceIPsFromOutgoingCtx(r.Context())
+
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
```

Comment:
There is no limit here. Hostile client can easily sends request with large body and crash Alertmanager. You can use `http.MaxBytesReader` to wrap `r.Body` and get error when too many bytes are read.